### PR TITLE
JP-2227: Making changes to properly create HDU from an ndarray.

### PR DESCRIPTION
**Pull request by kmacdonald-stsci**
Resolves JP-2227
Closes issue: https://github.com/spacetelescope/jwst/issues/6271

To support lazy loading in JWST data models a wrapper is used to wrap Numpy `ndarrays` called NDArrayType.  At the lower levels in the class hierarchy this causes problems during write.  When a data model is written an HDU needs to be created for the HDU list to be written out.  This can only be done properly if the `ndarray.dtype` is known.

For lazy loading this may not be known because the needed information is not in the NDArrayType wrapper, so an extra step must be taken to check if the type is NDArrayType, then ensure the `ndarray` gets loaded.  When this is done all the needed information for type checking on write is there and the round trip will succeed.

### CODE REVIEW

**Comment by nden** [comment]
Needs a change log

File: src/stdatamodels/fits_support.py
Comment (eslavich) [inline src/stdatamodels/fits_support.py:167]: This check needs to be higher up in the chain of methods.  The `hdu = hdu_type(value)` line just above will also fail if passed a NDArrayType, and there may be problems when updating an existing HDU that the check here doesn't address.

Comment (kmacdonald-stsci) [reply src/stdatamodels/fits_support.py]: Since `hdu_type` is only called in `_make_hdu` I kept the conditional in that function, but moved it to the beginning of the function.

**Review by hbushouse — APPROVED** [review]
Seems OK to me, but apparently my approval (or that of @stscieisenhamer) doesn't count for this repo.

File: src/stdatamodels/fits_support.py
Comment (eslavich) [inline src/stdatamodels/fits_support.py:159]: @kmacdonald-stsci did you check the update case?  I mean, when the HDU doesn't need to be created and the value is assigned to an existing HDU?

**Comment by hbushouse** [comment]
There's still one unanswered question from @eslavich and once that's resolved it needs approval from someone who's approved reviewer for this package (Jonathan and I are not, and hence our approvals don't count). @nden ?

File: src/stdatamodels/fits_support.py
Comment (kmacdonald-stsci) [reply src/stdatamodels/fits_support.py:159]: I checked this and it works fine.

**Review by eslavich — APPROVED** [review]
LGTM.  Ken and I chatted on Slack and I'm convinced that the update case doesn't have this problem.

**Comment by hbushouse** [comment]
Can someone with write access to `stdatamodels` merge this?  @nden @eslavich

**Comment by eslavich** [comment]
@hbushouse I also cursed you with admin privileges

**Comment by hbushouse** [comment]
> @hbushouse I also cursed you with admin privileges

A pox on both our houses?

**Pull request merged**
